src/open_company_web/actions.cljs

```diff
@@ -0,0 +1,91 @@
+(ns open-company-web.actions
+  (:require [medley.core :as med]
+            [open-company-web.dispatcher :refer [action dispatch!]]
+            [open-company-web.lib.utils :as utils]
+            [open-company-web.router :as router]
+            [open-company-web.caches :as cache]
+            [open-company-web.api :as api]))
+
+(defn- delay-remove-loading []
+  (.setTimeout js/window #(dispatch! [:clear-loading]) 100))
+
+;; ---- Generic Actions Dispatch
+;; This is a small generic abstraction to handle "actions".
+;; An `action` is a transformation on the app state.
+;; The return value of an action will be used as the new app-state.
+
+;; The extended multimethod `action` is defined in the dispatcher
+;; namespace to avoid cyclical dependencies between namespaces
+
+(defmethod action :default [db payload]
+  (js/console.warn "No handler defined for" (str (first payload)))
+  (js/console.log "Full event: " (pr-str payload))
+  db)
+
+(defmethod action :input [db [_ path value]]
+  (assoc-in db path value))
+
+(defmethod action :entry [db [_ {:keys [links]}]]
+  (if-let [co (med/find-first #(= (:rel %) "company") links)]
+    (router/nav! (str (:href co) "/dashboard"))
+    (if (med/find-first #(= (:rel %) "create-company") links)
+      (router/nav! "/create-company")))
+  db)
+
+(defmethod action :company-submit [db _]
+  (api/post-company (:company-editor db))
+  db)
+
+(defmethod action :company-created [db [_ body]]
+  (if (:links body)
+    (let [updated (utils/fix-sections body)
+          link (:href (med/find-first #(= "self" (:rel %)) (:links body)))]
+      (router/nav! (str link "/dashboard"))
+      (assoc db (keyword (:slug updated)) updated))
+    db))
+
+(defmethod action :new-section [db [_ body]]
+  (when body
```

Comment:
Because `dispatch-body` in the `api` namespace does fall back to an empty map in case of a unsuccessful response these `when` wraps don't do anything. I kept this behavior for now but probably in the future we'll want to clean this up somehow. (And report Xhr errors properly in the UI.)